gcloud-java-storage/src/main/java/com/google/gcloud/storage/CopyWriter.java

```diff
@@ -0,0 +1,269 @@
+/*
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.gcloud.storage;
+
+import static com.google.gcloud.RetryHelper.runWithRetries;
+
+import com.google.common.base.MoreObjects;
+import com.google.gcloud.Restorable;
+import com.google.gcloud.RestorableState;
+import com.google.gcloud.RetryHelper;
+import com.google.gcloud.spi.StorageRpc;
+import com.google.gcloud.spi.StorageRpc.RewriteRequest;
+import com.google.gcloud.spi.StorageRpc.RewriteResponse;
+
+import java.io.Serializable;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.Callable;
+
+/**
+ * Google Storage blob copy writer. This class holds the result of a copy request.
+ * If source and destination blobs do not share the same location or storage class more than one
+ * RPC request is needed to copy the blob otherwise one or more {@link #copyChunk()} calls are
+ * necessary to complete the copy.
+ *
+ * @see <a href="https://cloud.google.com/storage/docs/json_api/v1/objects/rewrite">Rewrite</a>
+ */
+public class CopyWriter implements Restorable<CopyWriter> {
+
+  private final StorageOptions serviceOptions;
+  private final StorageRpc storageRpc;
+  private RewriteResponse rewriteResponse;
+
+  CopyWriter(StorageOptions serviceOptions, RewriteResponse rewriteResponse) {
+    this.serviceOptions = serviceOptions;
+    this.rewriteResponse = rewriteResponse;
+    this.storageRpc = serviceOptions.rpc();
+  }
+
+  /**
+   * Returns the updated information for the just written blob. This method might block and issue
```

Comment:
Would be nice to indicate when it will block.
```
Returns the updated information for the written blob.
Calling this method when {@code isDone()} is {@code false} will 
block until all pending chunks are copied.
```